build/nested-cluster/start.sh

```diff
@@ -0,0 +1,121 @@
+#!/bin/bash
+
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+wait_for_docker ()
+{
+  # Wait for docker.
+  until docker version; do sleep 1 ;done
+}
+
+start_kubelet ()
+{
+  # Start the kubelet.
+  mkdir -p /etc/kubernetes/manifests
+  mkdir -p /etc/srv/kubernetes
+
+  # Change the kubelet to not fail with swap on.
+  cat > /etc/systemd/system/kubelet.service.d/kubeadm-20.conf << EOM
+[Service]
+Environment="KUBELET_EXTRA_ARGS=--fail-swap-on=false"
+EOM
+#  sed -i 's/ExecStart=\/usr\/bin\/kubelet/ExecStart=\/usr\/bin\/kubelet --fail-swap-on=false/' /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+  systemctl enable kubelet
+  systemctl start kubelet
+}
+
+start_worker ()
+{
+  wait_for_docker
+  start_kubelet
+
+  # Load docker images
+  docker load -i /kube-proxy-amd64.tar
+
+  # Start kubeadm.
+  /usr/bin/kubeadm join --token=abcdef.abcdefghijklmnop --discovery-token-unsafe-skip-ca-verification=true --ignore-preflight-errors=all 172.18.0.2:6443 2>&1
+}
+
+start_master ()
+{
+  wait_for_docker
+  mount --make-rshared /etc/kubernetes
+  start_kubelet
+
+  # Load the docker images
+  docker load -i /cloud-controller-manager-amd64.tar
+  docker load -i /kube-apiserver-amd64.tar
+  docker load -i /kube-controller-manager-amd64.tar
+  docker load -i /kube-proxy-amd64.tar
+  docker load -i /kube-scheduler-amd64.tar
```

Comment:
Are these stored in the docker image?

Reply:
They are in the final image with all the layers.

Reply:
I don't suppose it's particularly well optimised for time it takes to get a cluster up, cause that image would have to copied to each of the dockers, isn't it?
If so, I'd be keen to help finding a way to make it boot a cluster very fast.

Reply:
@errordeveloper this is absolutely the critical path. It's on the overlay file-system, which probably slows things down. Might we see significant performance improvements by squashing the layers?

Reply:
I am not sure what is the answer is, it'd be nice if we could just say "here is a bunch of layers that all dockers should just use". I wonder if there is a trick we could do on the filesystem layer, whereby wanted layers are there when docker daemon starts up.